File: src/apps/daemon/internal/networking/routing/policy_ruleset_router.go
package routing

import (
	"encoding/binary"
	"strings"
	"sync"
)

type PolicyVerdict string

const (
	PolicyVerdictDirect PolicyVerdict = "DIRECT"
	PolicyVerdictProxy  PolicyVerdict = "PROXY"
	PolicyVerdictReject PolicyVerdict = "REJECT"
)

type CidrRule struct {
	NetAddr uint32
	Mask    uint32
	Verdict PolicyVerdict
}

type KeywordRule struct {
	Keyword string
	Verdict PolicyVerdict
}

type PolicyRulesetRouter struct {
	mu            sync.RWMutex
	exactDomains  map[string]PolicyVerdict
	suffixDomains map[string]PolicyVerdict
	keywordRules  []KeywordRule
	cidrRules     []CidrRule
	cache         map[string]PolicyVerdict
	defaultPolicy PolicyVerdict
}

func NewPolicyRulesetRouter(defaultPolicy PolicyVerdict) *PolicyRulesetRouter {
	return &PolicyRulesetRouter{
		exactDomains:  make(map[string]PolicyVerdict),
		suffixDomains: make(map[string]PolicyVerdict),
		keywordRules:  make([]KeywordRule, 0),
		cidrRules:     make([]CidrRule, 0),
		cache:         make(map[string]PolicyVerdict),
		defaultPolicy: defaultPolicy,
	}
}

func (r *PolicyRulesetRouter) AddExactDomain(domain string, verdict PolicyVerdict) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.exactDomains[strings.ToLower(strings.TrimSpace(domain))] = verdict
}

func (r *PolicyRulesetRouter) AddSuffixDomain(suffix string, verdict PolicyVerdict) {
	r.mu.Lock()
	defer r.mu.Unlock()
	clean := strings.ToLower(strings.TrimSpace(suffix))
	clean = strings.TrimPrefix(clean, ".")
	r.suffixDomains[clean] = verdict
}

func (r *PolicyRulesetRouter) AddKeyword(keyword string, verdict PolicyVerdict) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.keywordRules = append(r.keywordRules, KeywordRule{
		Keyword: strings.ToLower(strings.TrimSpace(keyword)),
		Verdict: verdict,
	})
}

func (r *PolicyRulesetRouter) AddCidr(octets [4]byte, maskBits uint8, verdict PolicyVerdict) {
	r.mu.Lock()
	defer r.mu.Unlock()

	ipU32 := binary.BigEndian.Uint32(octets[:])
	var mask uint32
	if maskBits == 0 {
		mask = 0
	} else {
		mask = ^uint32(0) << (32 - maskBits)
	}

	r.cidrRules = append(r.cidrRules, CidrRule{
		NetAddr: ipU32 & mask,
		Mask:    mask,
		Verdict: verdict,
	})
}

func (r *PolicyRulesetRouter) ResolveDomain(domain string) PolicyVerdict {
	r.mu.Lock()
	defer r.mu.Unlock()

	clean := strings.ToLower(strings.TrimSpace(domain))
	if v, cached := r.cache[clean]; cached {
		return v
	}

	// 1. Exact match
	if v, exists := r.exactDomains[clean]; exists {
		r.cache[clean] = v
		return v
	}

	// 2. Suffix match
	for suffix, v := range r.suffixDomains {
		if clean == suffix || strings.HasSuffix(clean, "."+suffix) {
			r.cache[clean] = v
			return v
		}
	}

	// 3. Keyword match
	for _, kw := range r.keywordRules {
		if strings.Contains(clean, kw.Keyword) {
			r.cache[clean] = kw.Verdict
			return kw.Verdict
		}
	}

	r.cache[clean] = r.defaultPolicy
	return r.defaultPolicy
}

func (r *PolicyRulesetRouter) ResolveIP(octets [4]byte) PolicyVerdict {
	r.mu.RLock()
	defer r.mu.RUnlock()

	ipU32 := binary.BigEndian.Uint32(octets[:])
	for _, c := range r.cidrRules {
		if (ipU32 & c.Mask) == c.NetAddr {
			return c.Verdict
		}
	}
	return r.defaultPolicy
}

func (r *PolicyRulesetRouter) ClearCache() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.cache = make(map[string]PolicyVerdict)
}

File: src/packages/lumicore/src/routing/policy_ruleset_router.rs
//! # Policy Ruleset Router
//!
//! Evaluates policy rulesets across domain hierarchies, IP ranges, and user agents
//! to produce deterministic routing decisions (Direct, Proxy, Reject) with fast path caching.

use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PolicyVerdict {
    Direct,
    Proxy,
    Reject,
}

pub struct PolicyRulesetRouter {
    exact_domains: HashMap<String, PolicyVerdict>,
    suffix_domains: HashMap<String, PolicyVerdict>,
    keyword_rules: Vec<(String, PolicyVerdict)>,
    cidr_rules: Vec<([u8; 4], u8, PolicyVerdict)>,
    cache: HashMap<String, PolicyVerdict>,
    default_policy: PolicyVerdict,
}

impl PolicyRulesetRouter {
    pub fn new(default_policy: PolicyVerdict) -> Self {
        Self {
            exact_domains: HashMap::new(),
            suffix_domains: HashMap::new(),
            keyword_rules: Vec::new(),
            cidr_rules: Vec::new(),
            cache: HashMap::new(),
            default_policy,
        }
    }

    pub fn add_exact_domain(&mut self, domain: &str, verdict: PolicyVerdict) {
        self.exact_domains.insert(domain.trim().to_lowercase(), verdict);
    }

    pub fn add_suffix_domain(&mut self, suffix: &str, verdict: PolicyVerdict) {
        let mut clean = suffix.trim().to_lowercase();
        if clean.starts_with('.') {
            clean = clean[1..].to_string();
        }
        self.suffix_domains.insert(clean, verdict);
    }

    pub fn add_keyword(&mut self, keyword: &str, verdict: PolicyVerdict) {
        self.keyword_rules.push((keyword.trim().to_lowercase(), verdict));
    }

    pub fn add_cidr(&mut self, octets: [u8; 4], mask: u8, verdict: PolicyVerdict) {
        self.cidr_rules.push((octets, mask, verdict));
    }

    pub fn resolve_domain(&mut self, domain: &str) -> PolicyVerdict {
        let clean = domain.trim().to_lowercase();
        if let Some(&cached) = self.cache.get(&clean) {
            return cached;
        }

        // 1. Exact match
        if let Some(&v) = self.exact_domains.get(&clean) {
            self.cache.insert(clean, v);
            return v;
        }

        // 2. Suffix match
        for (suffix, &v) in &self.suffix_domains {
            if clean == *suffix || clean.ends_with(&format!(".{}", suffix)) {
                self.cache.insert(clean, v);
                return v;
            }
        }

        // 3. Keyword match
        for (kw, v) in &self.keyword_rules {
            if clean.contains(kw) {
                self.cache.insert(clean, *v);
                return *v;
            }
        }

        self.cache.insert(clean, self.default_policy);
        self.default_policy
    }

    pub fn resolve_ip(&self, ip: [u8; 4]) -> PolicyVerdict {
        let ip_u32 = u32::from_be_bytes(ip);
        for &(octets, mask_bits, verdict) in &self.cidr_rules {
            let net_u32 = u32::from_be_bytes(octets);
            let mask = if mask_bits == 0 {
                0u32
            } else {
                !0u32 << (32 - mask_bits)
            };
            if (ip_u32 & mask) == (net_u32 & mask) {
                return verdict;
            }
        }
        self.default_policy
    }

    pub fn clear_cache(&mut self) {
        self.cache.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_policy_ruleset_router() {
        let mut router = PolicyRulesetRouter::new(PolicyVerdict::Direct);
        router.add_exact_domain("ads.google.com", PolicyVerdict::Reject);
        router.add_suffix_domain("google.com", PolicyVerdict::Proxy);
        router.add_keyword("telegram", PolicyVerdict::Proxy);
        router.add_cidr([10, 0, 0, 0], 8, PolicyVerdict::Direct);

        assert_eq!(router.resolve_domain("ads.google.com"), PolicyVerdict::Reject);
        assert_eq!(router.resolve_domain("mail.google.com"), PolicyVerdict::Proxy);
        assert_eq!(router.resolve_domain("telegram.org"), PolicyVerdict::Proxy);
        assert_eq!(router.resolve_domain("wikipedia.org"), PolicyVerdict::Direct);

        assert_eq!(router.resolve_ip([10, 1, 2, 3]), PolicyVerdict::Direct);
        assert_eq!(router.resolve_ip([1, 1, 1, 1]), PolicyVerdict::Direct);
    }
}
